module/zfs/zfs_vfsops.c

```diff
@@ -1769,6 +1769,13 @@ zfs_remount(struct super_block *sb, int *flags, zfs_mnt_t *zm)
 	if (error)
 		return (error);
 
+	if ((dmu_objset_is_snapshot(zfsvfs->z_os) ||
+	    !spa_writeable(dmu_objset_spa(zfsvfs->z_os))) &&
+	    (!vfsp->vfs_readonly || !(*flags & MS_RDONLY))) {
```

Comment:
You need to remove the check on vfsp. As I said, we cannot rely on option string as MS_RDONLY is the only authoritative flag for readonly. It's perfectly normal to have MS_RDONLY and emtpy option string, and we need to treat that as readonly.